### ISSUE TITLE
Tabby server stops responding to requests after extended use

### ISSUE THREAD

**Issue by erfanium** [TabbyML/tabby#718]
I have deployed `tabbyml/tabby:0.5.3` with the arguments `serve --model TabbyML/StarCoder-3B --device cuda`. Initially, all functionalities work as expected, but after a period of using the model, the server stops responding to requests.

In this condition:
- The tabby process does not exit, and there are no errors or any other indications in the logs.
- Requests are never responded to by the tabby server, resulting in no server response.
- I have sufficient VRAM and am not running out of it.
- CPU/GPU Utilization is zero, which suggests there isn't an infinite loop.

I can usually reproduce this bug by sending many (e.g., 15) simultaneous `POST /v1/completions` requests to tabby.

Since there are no logs or errors, I cannot provide more information. Is there a verbose logging feature that I can enable?

### nvidia-smi
I have deployed `TabbyML/StarCoder-1B` and `TabbyML/StarCoder-3B` in the same machine
```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 545.29.02              Driver Version: 545.29.02    CUDA Version: 12.3     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3060        Off | 00000000:01:00.0 Off |                  N/A |
|  0%   56C    P8              16W / 170W |   6101MiB / 12288MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A     70768      C   /opt/tabby/bin/tabby                       2622MiB |
|    0   N/A  N/A     89712      C   /opt/tabby/bin/tabby                       3470MiB |
+---------------------------------------------------------------------------------------+
```

**Comment by erfanium** [comment]
I can sometimes reproduce this issue by this script
```py
import requests
import time

def send_requests_in_loop(base_url):
    # The data to be sent in the POST request
    data = {
        "language": "python",
        "segments": {
            "prefix": "# The point of life is",
            "suffix": ""
        }
    }

    # The full URL to which the POST request will be sent
    full_url = f'{base_url}/v1/completions'

    while True:
        # Record the start time
        start_time = time.time()

        # Send the POST request
        response = requests.post(full_url, json=data)

        # Record the end time
        end_time = time.time()

        # Calculate the response time
        response_time = end_time - start_time

        # Log the response time
        print(f"Response Time: {response_time} seconds")
        print(f"Response Body: {response.text}")

send_requests_in_loop('http://MY_TABBY_SERVER')
```

**Comment by erfanium** [comment]
Also is this okay that tabby uses too much CPU while tabby is idle?

![image](https://github.com/TabbyML/tabby/assets/47688578/3527e4a1-6f1b-442f-9fbb-1697340f987a)

**Comment by wsxiaoys** [comment]
Thanks for reporting the issue in a such informative way! I've reproduce the issue and working on a fix in #720

**Comment by wsxiaoys** [comment]
Hey - would you please test against the image tag `fix-tokio-deadlock-f532254` to see if it fixes the issue for you?

**Comment by erfanium** [comment]
@wsxiaoys Wow that was quick! Thanks

> Hey - would you please test against the image tag fix-tokio-deadlock-f532254 to see if it fixes the issue for you?

I will try that, but I've never build tabby from source code. it may take a while

**Comment by wsxiaoys** [comment]
> @wsxiaoys Wow that was quick! Thanks
> 
> > Hey - would you please test against the image tag fix-tokio-deadlock-f532254 to see if it fixes the issue for you?
> 
> I will try that, but I've never build tabby from source code. it may take a while

It's a docker image - no need for building it by your self: https://hub.docker.com/layers/tabbyml/tabby/fix-tokio-deadlock-f532254/images/sha256-e6914161c19535e18bf139468cb9ee9fb5f56cb752a716c26060a250c42834d0?context=explore

**Comment by erfanium** [comment]
@wsxiaoys 
It seems to be fixed, I couldn't reproduce that. thanks

but, the high CPU usage problem is still there. i couldn't remember i have this issue with older versions of tabby (2 months ago). is it expected? should i create new issue for that?

**Comment by wsxiaoys** [comment]
Could you please test it again with the `fix-tokio-deadlock-ece5f3c` image? This should resolve the cpu usage caused by busy waiting.

For the context in release 0.5, we transitioned the CUDA inference backend from ctranslate2 to llama.cpp for improved flexibility and performance.
However, this change introduced several issues that were not covered in our release tests. We appreciate your feedback and apologize for any inconvenience this may have caused you.

**Comment by wsxiaoys** [comment]
#720 will be released in 0.5.4

**Comment by wsxiaoys** [comment]
Done in https://github.com/TabbyML/tabby/releases/tag/v0.5.4
